keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -229,7 +229,7 @@ kind: CustomResourceDefinition
         spec:
           x-kubernetes-validations: 
             - rule: "self.minReplicas <= self.maxReplicas"
-              message: "minReplicas cannot be larger than maxReplicas"
+              messageExpression: "'minReplicas ({}) cannot be larger than maxReplicas ({})'.format([self.minReplicas, self.maxReplicas])"
```

Comment:
how does evaluation of another expression interact with cost limits (static or runtime)? if these are included in the cost limits, what is the behavior if the rule completes within runtime cost limits and the messageExpression does not?

Reply:
what variables are available to the messageExpression? if `rule` does not use `oldSelf`, is `messageExpression` allowed to use it?

Reply:
For runtime cost limits: `messageExpression` can fail due to runtime costs - if it does, the default error message is used instead and the fact that `messageExpression` went over the limit is logged. Whether or not the action is admitted in that instance depends on the failure policy.

Good question about variable availability as well. I think the simplest/least-surprising behavior is to limit variable usage to whatever the rule has access to/uses (so if `oldSelf` is not used by the rule it isn't available). An exception to that would be variables as defined as a part of https://github.com/kubernetes/enhancements/pull/3669, I think.

For behavior post-runtime cost and variable availability I had had both as a part of https://github.com/kubernetes/enhancements/pull/3736, so I'll go ahead and add both in so both PRs are in sync. (Keeping this unresolved in the immediate short-term since I'm not sure if that answers your questions, though.)

Reply:
> if it does, the default error message is used instead

I'm not sure what this means... the generated message using only the rule as if no `message` or `messageExpression` was provided? That is not always suitable for user display, which is why we require `message` if `rule` contains newlines:

```go
if hasNewlines(trimmedRule) && len(trimmedMsg) == 0 {
  allErrs.SchemaErrors = append(allErrs.SchemaErrors, field.Required(fldPath.Child("x-kubernetes-validations").Index(i).Child("message"), "message must be specified if rule contains line breaks"))
}
```

Reply:
another question: are we able to validate the format string and params, so that these would be flagged at validation time, not at runtime:
* `'foo'.format([self.minReplicas])` // more params than tokens
* `'foo {}'.format([])` // more tokens than params

Reply:
something else to consider is how a 1.26 API server would see and validate a rule that specified a messageExpression and no message...

a 1.26 API server would require something like this (because `rule` has newlines):
```yaml
- rule: "self.valid &&\n!self.invalid"
  message: "must be valid and not invalid"
```

would a 1.27 API server allow something like this?
```yaml
- rule: "self.valid &&\n!self.invalid"
  messageExpression: "'valid must be true (was {}), and invalid must be false (was {})'.format([self.valid, self.invalid])"
```

if so, that would be seen like this by a 1.26 server and would fail validation:
```yaml
- rule: "self.valid &&\n!self.invalid"
```

Reply:
>  I'm not sure what this means... the generated message using only the rule as if no message or messageExpression was provided? That is not always suitable for user display, which is why we require message if rule contains newlines:

Very good point - I think requiring a backup `message` field to be present everywhere `messageExpression` would be used is a bit excessive; maybe we could output `rule` truncated up to the first newline in addition to outputting the source code of `messageExpression`  in those cases? I don't think it's ideal, but it should still give users something to search for.

> another question: are we able to validate the format string and params, so that these would be flagged at validation time, not at runtime:

The `format` implementation for cel-go will flag those as compile-time errors if (and only if) the formatting string is a constant, so we should be fine on that front.

> something else to consider is how a 1.26 API server would see and validate a rule that specified a messageExpression and no message...

Good point - I'll adjust it such that `message` and `messageExpression` are not mutually-exclusive (something I believe @cici37 originally requested - terribly sorry for not updating this copy of the messageExpression proposal everyone).

Reply:
`message` is [optional](https://github.com/kubernetes/kubernetes/blob/200220ccefad096c5d539037c5bd663feb1210e7/staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go#L180-L182).  I'm OK with not having the fields be mutually exclusive, but we should define the behavior carefully. I would expect `messageExpression` to take precedence if set and if it returns a string, but for `message` to be used otherwise..

Reply:
+1 to pre-compiling format. This was discussed during design and my understanding is that it is in scope.

I expect the rules for `format(fmtString, args)` to be:

- If `fmtString` is a string constant (almost always the case), verify the string parses
- If the `args` are defined as an array literal (almost always the case), validate carnality against `fmtString`
- If the CEL expression is compiled, verify that the types of all `args` match the expected types of the directives in `fmtString`

Reply:
Right, the implementation will have all 3 forms of checks.

Reply:
Can we merge the above into the KEP and resolve this one?